### ISSUE TITLE
Refresh OAuth 2 Access token with Refresh Token failed

### ISSUE THREAD

**Issue by mishabachkur** [intuit/QuickBooks-V3-PHP-SDK#324]
I often get this error - Refresh OAuth 2 Access token with Refresh Token failed

**Comment by bsivalingam** [comment]
@mishabachkur This error typically occurs when you have an incorrect configuration for clientId, clientSecret etc. Can you please check you have configure the keys accurately?

**Comment by coresh** [comment]
### The issue: confirmed

The reason:

With the same credentials:
```
        $companyInfo = $dataService->getCompanyInfo();
        return json_encode($companyInfo, true);
```
worked properly

But,

```
        $OAuth2LoginHelper = $dataService->getOAuth2LoginHelper();
        $accessTokenObj    = $OAuth2LoginHelper->exchangeAuthorizationCodeForToken(
            self::QUICKBOOKS_AUTHORIZATION_CODE,
            self::QUICKBOOKS_REALM_ID
        );
```
Provided with an issue:

```[RuntimeException]                                                                                
  Cron-job "socalgraph_quickbooks" threw exception QuickBooksOnline\API\Exception\ServiceException  
                                                                                                    

                                                                                                                               
  [QuickBooksOnline\API\Exception\ServiceException]                                                                            
  Exchange Authorization Code for Access Token failed. Body: [{"error_description":"Token invalid","error":"invalid_grant"}].  
```

**Comment by bsivalingam** [comment]
@coresh That is very interesting find. Thanks for persisting on the issue. Would you be able to send me the raw request and response log for both the calls? I can bring it to my downstream team.

**Comment by coresh** [comment]
```
vendor/quickbooks/v3-php-sdk/src/Core/OAuth/OAuth2/OAuth2LoginHelper.php

public function exchangeAuthorizationCodeForToken($code, $realmID){
...
var_dump(http_build_query($parameters));
var_dump($intuitResponse->getBody());
var_dump($intuitResponse->getHeaders());
...
```
Result:

```string(171) "grant_type=authorization_code&code=TheRealauthorizationCode&redirect_uri=https%3A%2F%2Fdeveloper.intuit.com%2Fv2%2FOAuth2Playground%2FRedirectUrl"
eveloper.intuit.com%2Fv2%2FOAuth2Playground%2FRedirectUrl"

string(61) "{"error_description":"Token invalid","error":"invalid_grant"}"

array(10) {
  ["date"]=>
  string(29) "Mon, 13 Jul 2020 16:24:34 GMT"
  ["content-type"]=>
  string(30) "application/json;charset=utf-8"
  ["content-length"]=>
  string(2) "61"
  ["server"]=>
  string(5) "nginx"
  ["strict-transport-security"]=>
  string(16) "max-age=15552000"
  ["intuit_tid"]=>
  string(35) "1-5f0c8ac2-94777b4a0a0f25100de4b6f2"
  ["x-spanid"]=>
  string(38) "6f58449e-aa9e-4cf7-b640-8ec0744cc449:1"
  ["x-amzn-trace-id"]=>
  string(81) "Self=1-5f0c8ac2-c63f2990c703c9bc452205fc;Root=1-5f0c8ac2-94777b4a0a0f25100de4b6f2"
  ["cache-control"]=>
  string(18) "no-cache, no-store"
  ["pragma"]=>
  string(8) "no-cache"
}
```

**Comment by bsivalingam** [comment]
@coresh Ah I see what you are saying. These calls use different auth headers. The auth code call uses `Basic XXXX` where `XXXX` is a base 64 encoded value of client id and secret. The getUserInfo call is a Bearer token call, where you use the accessToken which you had last obtained. I would still suggest checking the value for clientId that is being passed in the request. Since the client information are secure, you could choose to create a support ticket and we can help you there.

**Comment by coresh** [comment]
Re : `I would still suggest checking the value for clientId that is being passed in the request.`

```clientID``` -> it's Ok.

The reason:

call:

```$dataService->getCompanyInfo()```
 
returns properly output

Please check concerning to the `Production` environment.

Thank you

**Comment by bsivalingam** [comment]
@coresh can you please elaborate what you mean by "Please check concerning to the Production environment."?

**Comment by daraul** [comment]
I am getting the same error. I have found that I get a new refresh token *before* 24 hours have elapsed since I received the old one. My application is configured to save the new token every time, but the new token *does not work*. If I log the old token, and use that to call the refresh token endpoint again it works fine, and I get the old token back. I do believe there might be an issue on your guy's end. 

Checking my error logs I notice that I can't re-authenticate my server about 8 hours after doing so manually, leading me to believe that that's when you guys return a new refresh token to me. 

My client ID, and secret are both correct, as if I go to my server and simply change the refresh token back to what it was before the automated process I can retrieve a new access token, but a new refresh token is *not* returned. 

This is confusing so I'll go through step by step what I did, and how I think things went on the server:

1. use the playground to retrieve an access and refresh token
1. log both tokens at the time i receive them in my issue tracker 
1. save those on the server, and begin the re-authentication job (this runs every 15 minutes to retrieve a new access token)
1. *8 hours later, on the server*: quickbooks returns a new refresh token, earlier than expected, but it's saved anyway, as recommended by the documentation at point 3 [here](https://intuit.github.io/QuickBooks-V3-PHP-SDK/authorization.html#oauth-2-0-vs-1-0a-in-quickbooks-online)
1. that new token does not work, leading to the `invalid_grant` error from quickbooks
1. I notice the error (at the time of writing, roughly 22 hours after I initiated step 1), log into the server, with the *old refresh token*, which I logged earlier, and replace the *new refresh token I got from quickbooks* with the old one
1. call the refresh_token endpoint, with the old token, which returns a new access_token as expected, and *the old refresh token*. 

I think that around 8 hours after I initiated step 1, you guys return a new refresh token to me.

**Comment by npapratovic** [comment]
In short, the recommended workflow for all apps is:
Step 1. If the Access Token fails, use the current Refresh Token to request new Access and Refresh Tokens.
Step 2. Store BOTH the Access and Refresh Tokens that are returned.
Step 3. Use the new Access Token for making QuickBooks Online API calls.
Step 4. When the Access Token fails in 1 hour, go to Step 1.

To solve this error: "Refresh OAuth 2 Access token with Refresh Token failed" in my Laravel application I had this approach:

1) store config data in .env files, store tokens in database; fetch them on every DataService object creation, and recreate tokens on the fly and save them again in database: 

```

use QuickBooksOnline\API\Core\OAuth\OAuth2\OAuth2LoginHelper;
use QuickBooksOnline\API\DataService\DataService;
use QuickBooksOnline\API\Exception\SdkException;
use QuickBooksOnline\API\Exception\ServiceException;


    public function getTokensFromDataBase()
    {
        $config = config('quickbooks');
        $quickbook = \App\Models\QuickBook::where('realm_id', $config['realm_id'])->first();
        return [
            'access_token' => $quickbook->access_token,
            'refresh_token' => $quickbook->refresh_token,
            'x_refresh_token_expires_in' => $quickbook->x_refresh_token_expires_in,
        ];
    }




    public function getTokens()
    {
    	// step 0 fetch latest tokens from db
        $tokens_from_db = $this->getTokensFromDataBase();

        // step 1: get config data, and get refresh token
        $config = config('quickbooks');
        $dataService = DataService::Configure([
            'auth_mode' => 'oauth2',
            'ClientID' => $config['client_id'],
            'ClientSecret' => $config['client_secret'],
            'RedirectURI' => $config['redirect_uri'],
            'scope' => 'com.intuit.quickbooks.accounting',
            'baseUrl' => $config['base_url'],
            // Get the refresh token from session or database
            'refreshTokenKey' => $tokens_from_db['refresh_token'],
            'QBORealmId' => $config['realm_id'],
        ]);

        // step 2: refresh access token using the refresh token
        // The first parameter of OAuth2LoginHelper is the ClientID, the second parameter is the client Secret
        $oauth2LoginHelper = new OAuth2LoginHelper($config['client_id'], $config['client_secret']);
        $accessToken = $oauth2LoginHelper->refreshAccessTokenWithRefreshToken($tokens_from_db['refresh_token']);
        $dataService->updateOAuth2Token($accessToken);
        // step 3: save the new tokens in the database

        $token_in_db = \App\Models\QuickBook::where('realm_id', $config['realm_id'])->first();
        $token_in_db->access_token = $accessToken->getAccessToken();
        $token_in_db->refresh_token = $accessToken->getRefreshToken();
        $token_in_db->x_refresh_token_expires_in = $accessToken->getRefreshTokenExpiresAt();
        $token_in_db->save();

        return [
            'access_token' => $accessToken->getAccessToken(),
            'refresh_token' => $accessToken->getRefreshToken(),
            'x_refresh_token_expires_in' => $accessToken->getRefreshTokenExpiresAt(),
            'expires_in' => $accessToken->getAccessTokenExpiresAt(),
        ];
    }

```
2) In your controller where you wwant to store invoice in QB, fetch tokens by calling method getTokens() and set DataService object again:

```
		$tokens = $this->getTokens();
 
        $config = config('quickbooks');

        $dataService = DataService::Configure([
            'auth_mode' => 'oauth2',
            'ClientID' => $config['client_id'],
            'ClientSecret' => $config['client_secret'],
            'RedirectURI' => $config['redirect_uri'],
            'accessTokenKey' => $tokens['access_token'],
            'refreshTokenKey' => $tokens['refresh_token'],
            'QBORealmID' => $config['realm_id'],
            'baseUrl' => $config['base_url'],
        ]);

        $dataService->throwExceptionOnError(true);
```


3) now, when DataSerice object is set up properly, you can follow instructions to save or ftech data to / from QB: 

```

	....


        $invoice = Invoice::create([
            "Line" => $lines,
            "CustomerRef" => [
                "value" => $customer->Id,
                "name" => $customer->name,
            ],
            "BillEmail" => [
                "Address" => $customer->email,
            ],
        ]);

        $response_from_qb = $dataService->Add($invoice);

        $xmlResponseQB = XmlObjectSerializer::getPostXmlFromArbitraryEntity($response_from_qb, $urlResource);

        // use this to debug response from QB
        echo "Invoice Created Id={$response_from_qb->Id}. Reconstructed response body:\n\n";
        echo $xmlResponseQB;


	....

```

4) problem with tokens is that they expire - access token is valid for 60 minutes, and by using approach above, if your users do not perform query to the QB, access tokens in your db will expire

My approach was to create Laravel command, and set up cron job which will run every 10 minutes and update tokens in db by itself: 

```
<?php

namespace App\Console\Commands;

use Illuminate\Console\Command;
use QuickBooksOnline\API\Core\OAuth\OAuth2\OAuth2LoginHelper;
use QuickBooksOnline\API\DataService\DataService;
use QuickBooksOnline\API\Exception\SdkException;
use QuickBooksOnline\API\Exception\ServiceException;

class refreshQBToken extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'app:refresh-q-b-token';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'This command refreshes the QuickBooks token.';

    /**
     * Execute the console command.
     * @throws SdkException
     * @throws ServiceException
     */
    public function handle()
    { 

        $config = config('quickbooks');
        $token_in_db = \App\Models\QuickBook::where('realm_id', $config['realm_id'])->first();
        $dataService = DataService::Configure([
            'auth_mode' => 'oauth2',
            'ClientID' => $config['client_id'],
            'ClientSecret' => $config['client_secret'],
            'RedirectURI' => $config['redirect_uri'],
            'scope' => 'com.intuit.quickbooks.accounting',
            'baseUrl' => $config['base_url'],
            // Get the refresh token from session or database
            'refreshTokenKey' => $token_in_db->refresh_token,
            'QBORealmId' => $config['realm_id'],
        ]);
        $oauth2LoginHelper = new OAuth2LoginHelper($config['client_id'], $config['client_secret']);
        $accessToken = $oauth2LoginHelper->refreshAccessTokenWithRefreshToken($token_in_db->refresh_token);
        $dataService->updateOAuth2Token($accessToken);
        //save the new tokens in the database
        $token_in_db->access_token = $accessToken->getAccessToken();
        $token_in_db->refresh_token = $accessToken->getRefreshToken();
        $token_in_db->x_refresh_token_expires_in = $accessToken->getRefreshTokenExpiresAt();
        $token_in_db->save();

        $this->info('Successfully refreshed QB tokens.');
    }
}
```



Thats how I managed to solve problem with tokens expiry

**Comment by daraul** [comment]
> @daraul Have you figure out how to solve this?

Sorry for the very late response, @radenkozec -- no, I never figured out a workaround, and I've since left that job so it's someone else's problem, now 😆 

Maybe the recommendation above will work.

**Comment by jainnys** [comment]
Here is the explanation to this issue:
https://help.developer.intuit.com/s/article/Validity-of-Refresh-Token

`On day 2, a new Refresh Token is returned, and Refresh Token A expires. Refresh token A is no longer valid for making the Refresh Token API call to get a new access token. The Refresh Token API call will throw an “invalid_grant” error if you attempt to use it.
 For simplicity, always store and use the LATEST Refresh token returned.

In short, the recommended workflow for all apps is:
Step 1. If the Access Token fails, use the current Refresh Token to request new Access and Refresh Tokens.
Step 2. Store BOTH the Access and Refresh Tokens that are returned.
Step 3. Use the new Access Token for making QuickBooks Online API calls.
Step 4. When the Access Token fails in 1 hour, go to Step 1.
 `

**Comment by siddharthb2b** [comment]
@jainnys when I am trying to generate the new refresh token using the current refresh token, it returns same value on 1st day, on day 2 after 24 hours, it throws an error "invalid_grant"

**Comment by pratushraj** [comment]
Same issue with me... Any solutions

**Comment by ghost** [comment]
Hi there! 

I work on Nango (open-source OAuth + token refresh): [Nango](https://github.com/NangoHQ/nango)
From what we see across lots of APIs, refresh tokens can get revoked even when everything looks “correct”, so it’s often not worth over-optimising for a single root cause. 

The pragmatic fix is solid revocation detection plus a smooth re-connect flow. The most common implementation gotchas are: 
**(1)** refresh race conditions (parallel refresh attempts)
**(2)** not saving the latest refresh token (rotation)
**(3)** leaving tokens idle too long (some providers revoke if unused).

Really hope helps you skip a bit of extra digging!

**Comment by Janishia** [comment]
Hi all — to clarify the root cause behind the invalid_grant / "refresh token not found" errors reported here:

The Refresh Token API call is used to get a new 1-hour Access Token when the previous one expires. Per the recent policy change, refresh tokens now have a maximum validity period of five years — so you shouldn't need to re-prompt users for consent until then. However, this can shift about a day later, or at the next Refresh Token API call, whichever is later. Importantly: if a new Refresh Token is returned, the previous one immediately expires.

This explains the specific symptoms reported in this thread:

- @daraul, @siddharthb2b, @pratushraj — if you log in again within 24 hours (via createBearerTokenFromAuthorizationCode), a new refresh token is generated. If your app is still holding onto and using the old refresh token, you'll get invalid_grant/"refresh token not found" — because it's already been superseded.
- Similarly, if a logout (revokeConsentByClientAndUser) happens within 24 hours of the last refresh, a new refresh token is generated on the next login, and the old one is invalidated the same way.

The fix (matches @npapratovic's approach above): always persist and use the latest token pair, and reset both tokens on every refresh — not just the access token:
1. Test the access token.
2. If it fails, use the current refresh token to request new tokens.
3. Store both the new access token and the new refresh token (encrypted at rest).
4. Use the new access token to connect.
5. When the access token expires in an hour, go back to step 2.
